src/util.tsx

```diff
@@ -93,11 +89,10 @@ export function calcDropPosition(
   dropAllowed: boolean,
 } {
   const { clientX, clientY } = event;
-  const { top, height } = (event.target as HTMLElement).getBoundingClientRect();
+  const { top, height, left, right } = (event.target as HTMLElement).getBoundingClientRect();
   // optional chain for testing
-  const horizontalMouseOffset = (direction === 'rtl' ? -1 : 1) * ((startMousePosition?.x || 0) - clientX);
+  const horizontalMouseOffset = (direction === 'rtl' ? -1 : 1) * ((((right + left) / 2) || 0) - clientX);
```

Comment:
We used to implement offset like this but find the user experience is not good. Can we just memorized startMousePosition in dragNode's DOM node with a special prop?

If we find the dragging node has the prop and currently the drop tree has no startMousePosition, we can set drop tree's startMousePosition from the dragging dom node. Hence we can ensure the same dnd experience for both tree.

Reply:
But that would never produce `horizontalMouseOffset < 0` if you drag from right to left, and never `horizontalMouseOffset > 0` if you drag from left to right.
This was my first approach.

How about we keep the startMousePosition approach but then ´if startMousePosition === null´ use the bbox approach as fallback to support DnD between trees?

DnD between trees is an important functionality for some projects (see also https://github.com/react-component/tree/pull/338) and maybe we could accept a slight compromise in user experience for this specific use case?

Reply:
I committed the fallback solution now

Reply:
> But that would never produce `horizontalMouseOffset < 0` if you drag from right to left, and never `horizontalMouseOffset > 0` if you drag from left to right.
> This was my first approach.
> 
> How about we keep the startMousePosition approach but then ´if startMousePosition === null´ use the bbox approach as fallback to support DnD between trees?
> 
> DnD between trees is an important functionality for some projects (see also #338) and maybe we could accept a slight compromise in user experience for this specific use case?

You are right. Memorize `startMousePosition` is not the correct way. Bbox is a reasonable solution.

I've another proposal to discuss: We memorized the `mouseStartPosition`'s offset to the left edge of the dragging start tree. We can use this value to restore a valid `startMousePositon.x` in another tree (by adding bbox.left of another tree).

Compared with bbox only solution, I think it can keep the dragging experience more consistantly.

Reply:
@thomasstjerne

Reply:
@07akioni I fixed the bug. 

> I've another proposal to discuss: We memorized the mouseStartPosition's offset to the left edge of the dragging start tree. We can use this value to restore a valid startMousePositon.x in another tree (by adding bbox.left of another tree).

I can look into this, but I have little time these days. Would you consider merging the current bbox fallback solution, and then I´ll do a new PR if I succeed with your proposal?

Reply:
Sorry for can't merging the current changes. The code still need to be modified.

The main reason is maybe it's not necessary to expose `cleanDragState` to the user (It is a big change and can't be reverted).